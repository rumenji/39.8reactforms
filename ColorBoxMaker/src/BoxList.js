import React, { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import { v4 as uuid } from "uuid";

function BoxList(){
    const [boxes, setBoxes] = useState([]);

    const addBox = box => {
        const newBox = {...box, id: uuid()};
        setBoxes(boxes => [...boxes, newBox])
    }

    const removeBox = idx => {
        const boxesCopy = [...boxes];
        setBoxes(boxes => boxesCopy.filter(box => box.id !== idx))
    }

    return (
        <div className="BoxList">
            <NewBoxForm addBox={addBox} />
            {boxes.map((box) => (<Box style={box.style} remove={() => removeBox(box.id)} testId={box.id}/>))}
        </div>
    )
}

export default BoxList;