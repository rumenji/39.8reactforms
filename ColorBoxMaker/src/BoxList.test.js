import React from "react";
import { render, fireEvent } from '@testing-library/react';
import BoxList from "./BoxList";

it("renders without crashing", function(){
    render(<BoxList />)
})

it("matches snapshot", function(){
    const {asFragment} = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
})

it("can add new boxes", function(){
    const {getByLabelText, queryByText, queryByTestId} = render(<BoxList />);

    const colorInput = getByLabelText("Pick a color:");
    const heightInput = getByLabelText("Heigth in px:");
    const widthInput = getByLabelText("Width in px:");
    const submitBtn = queryByText("Add box")

    fireEvent.change(colorInput, {target: {value: "blue"}});
    fireEvent.change(heightInput, {target: {value: "100px"}});
    fireEvent.change(widthInput, {target: {value: "100px"}});

    fireEvent.click(submitBtn);

    expect(queryByTestId("0")).toBeInTheDocument();
})

it("can remove boxes", function(){
    const {getByLabelText, queryByText, queryByTestId} = render(<BoxList />);

    const colorInput = getByLabelText("Pick a color:");
    const heightInput = getByLabelText("Heigth in px:");
    const widthInput = getByLabelText("Width in px:");
    const submitBtn = queryByText("Add box")

    fireEvent.change(colorInput, {target: {value: "blue"}});
    fireEvent.change(heightInput, {target: {value: "100px"}});
    fireEvent.change(widthInput, {target: {value: "100px"}});

    fireEvent.click(submitBtn);

    expect(queryByTestId("0")).toBeInTheDocument();

    const removeBtn = queryByText("X");
    fireEvent.click(removeBtn);

    expect(queryByTestId("0")).not.toBeInTheDocument();

})