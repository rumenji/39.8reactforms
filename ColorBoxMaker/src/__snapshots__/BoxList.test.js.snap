// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <form>
      <label
        for="color"
      >
        Pick a color:
      </label>
      <input
        id="color"
        name="backgroundColor"
        placeholder="color"
        type="text"
        value=""
      />
      <label
        for="heigth"
      >
        Heigth in px:
      </label>
      <input
        id="heigth"
        name="height"
        placeholder="height"
        type="text"
        value=""
      />
      <label
        for="width"
      >
        Width in px:
      </label>
      <input
        id="width"
        name="width"
        placeholder="width"
        type="text"
        value=""
      />
      <button>
        Add box
      </button>
    </form>
  </div>
</DocumentFragment>
`;
